### PR TITLE
Adds Support for Push Authentication

### PR DESCRIPTION
**Pull request by roundhill**
screen shot 2015-06-01 at 1 00 27 pm

A new way for two step authentication to WordPress.com! If you sign in via a web browser to WordPress.com and you are using a WP app, you will receive a push to your device that opens to an alert dialog that allows you to approve or ignore the request.

To test this, you'll need to:
1. Use an Automattician account with two step authentication enabled.
2. Set the build number in `WordPress/build.gradle` to 4.2 or higher and rebuild to your device.
3. Finally, sign in to WordPress.com from a web browser.

Note: There's currently a bug where the location in the push notification isn't sent properly. This will be a server-side fix.

### CODE REVIEW

**Comment by nbradbury** [comment]
@roundhill I went to login using my 2fa password, and my phone correctly received the push and displayed the "Verify login" screen. However, rather than approve it I chose to ignore it,  then on the web I chose to use a verification code (which I never entered).

Now every time I try to login I'm asked for a verification code - I never see anything about an approval being sent to my phone any more. So I'm unable to test this further.

File: WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java [hunk 1]
```diff
@@ -149,6 +150,24 @@ protected void onNewIntent(Intent intent) {
     private void launchWithNoteId() {
         if (isFinishing() || getIntent() == null) return;
 
+        // Check for push authorization request
+        if (getIntent().hasExtra(NotificationsUtils.ARG_PUSH_AUTH_TOKEN)) {
+            Bundle extras = getIntent().getExtras();
+            String token = extras.getString(NotificationsUtils.ARG_PUSH_AUTH_TOKEN, "");
+            String title = extras.getString(NotificationsUtils.ARG_PUSH_AUTH_TITLE, "");
+            String message = extras.getString(NotificationsUtils.ARG_PUSH_AUTH_MESSAGE, "");
+            long expires = extras.getLong(NotificationsUtils.ARG_PUSH_AUTH_EXPIRES, 0);
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java]: Should the default expiration if none is passed both here and [here](https://github.com/wordpress-mobile/WordPress-Android/pull/2788/files#diff-6493e140350cb7e12c1195918fd6ebd8R284) really be `0`?

My concern is that for some reason the backend changes and `expires` is no longer passed, which would mean the app would always show the toast telling the user they took too long.

File: WordPress/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -558,6 +558,8 @@
     <string name="older_last_week">Older than a week</string>
     <string name="older_month">Older than a month</string>
     <string name="error_notification_open">Could not open notification</string>
+    <string name="ignore">Ignore</string>
+    <string name="push_auth_expired">The login request has expired. Log in to WordPress.com to try again.</string>
```
Comment (nbradbury) [inline WordPress/src/main/res/values/strings.xml]: We use different wordage for logging in ("login" and "log in" and "sign in") in different places in the app. The most prevalent is "sign in" - can we use that here, too?

**Comment by roundhill** [comment]
> However, rather than approve it I chose to ignore it

I believe you can get one of these pushes every 15 minutes. Can you try again now?

File: WordPress/src/main/res/values/strings.xml (hunk 2 above)
Comment (roundhill) [inline WordPress/src/main/res/values/strings.xml]: Sure, sounds good. How about:

`The request has expired. Sign in to WordPress.com to try again."

Interestingly enough, WordPress.com uses 'Log In' but has a checkbox that says 'Stay signed in'
screen shot 2015-06-03 at 9 42 31 am

**Comment by nbradbury** [comment]
> I believe you can get one of these pushes every 15 minutes. Can you try again now?

Ah, got it - it's working now (and it works well).

File: WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java (hunk 1 above)
Comment (roundhill) [inline WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java]: Good catch, it should still show the prompt if we don't receive an expiry date. Fixed in 577b9bcf44d73012978b96f888f297b37f6c042c

**Comment by nbradbury** [comment]
:shipit:

**Pull request merged**